File: js/View2d.js
export class View2d {

    constructor(model, canvas2d) {
        this.model = model;
        this.canvas2d = canvas2d;

        // Resize
        window.addEventListener('resize', () => {
            this.fit();
            this.drawModel();
        });
    }

    // Fit to show all the model in the view
    fit() {
        // Model
        const {xMin, yMin, xMax, yMax} = this.model.get2DBounds();
        const modelWidth = xMax - xMin;
        const modelHeight = yMax - yMin;
        // <div> containing Canvas
        const rect = this.canvas2d.getBoundingClientRect();
        const viewWidth = rect.width;
        const viewHeight = rect.height;
        // Resize canvas2d to fit inside <div>
        this.canvas2d.width = viewWidth - 2 * this.canvas2d.clientLeft;
        this.canvas2d.height = viewHeight - 2 * this.canvas2d.clientTop;
        // Compute scale to fit
        const scale = Math.min(viewWidth / modelWidth, viewHeight / modelHeight) / 1.2;
        // Compute offset to center drawing
        const xOffset = viewWidth / 2;
        const yOffset = viewHeight / 2;

        const context2d = this.canvas2d.getContext('2d');
        context2d.setTransform(scale, 0, 0, scale, xOffset, yOffset);
    }

    // Draw points
    drawPoints(points) {
        const context2d = this.canvas2d.getContext('2d');
        context2d.font = '18px serif';
        points.forEach(p => {
            const xc = p.xf;
            const yc = -p.yf;
            context2d.beginPath();
            context2d.arc(xc, yc, p.hover ? 12 : 8, 0, 2 * Math.PI);
            context2d.fillStyle = p.select === 1 ? 'red' : p.select === 2 ? 'orange' : p.hover ? 'blue' : 'skyblue';
            context2d.fill();

            // label
            context2d.fillStyle = p.hover ? 'white' : 'black';
            const label = String(this.model.points.indexOf(p));
            context2d.fillText(label, xc - 4 * label.length, yc + 5);
        });
    }

    // Draw segments
    drawSegments(segments) {
        const context2d = this.canvas2d.getContext('2d');
        context2d.font = '18px serif';
        context2d.lineWidth = 4;

        segments.forEach(s => {
            const [xf1, yf1, xf2, yf2] = [s.p1.xf, -s.p1.yf, s.p2.xf, -s.p2.yf];
            const [xc, yc] = [(xf1 + xf2) / 2, (yf1 + yf2) / 2];

            context2d.lineWidth = s.hover ? 6 : 3;
            context2d.beginPath();
            context2d.moveTo(xf1, yf1);
            context2d.lineTo(xf2, yf2);
            context2d.strokeStyle = s.select === 1 ? 'red' : s.select === 2 ? 'orange' : s.hover ? 'blue' : 'skyblue';
            context2d.stroke();
            // Add an arrow on p2
            const arrowLength = 20;
            const arrowAngle = Math.PI / 7;
            const angle = Math.atan2(yf2 - yf1, xf2 - xf1);
            const [arrowX1, arrowY1] = [xf2 - arrowLength * Math.cos(angle + arrowAngle), yf2 - arrowLength * Math.sin(angle + arrowAngle)];
            const [arrowX2, arrowY2] = [xf2 - arrowLength * Math.cos(angle - arrowAngle), yf2 - arrowLength * Math.sin(angle - arrowAngle)];
            context2d.beginPath();
            context2d.moveTo(xf2, yf2);
            context2d.lineTo(arrowX1, arrowY1);
            context2d.lineTo(arrowX2, arrowY2);
            context2d.closePath();
            context2d.fillStyle = s.select === 1 ? 'red' : s.select === 2 ? 'orange' : s.hover ? 'blue' : 'skyblue';
            context2d.fill();

            // Circle with color for selected, bigger for hovered
            context2d.beginPath();
            context2d.arc(xc, yc, s.hover ? 12 : 8, 0, 2 * Math.PI);
            context2d.fillStyle = s.select === 1 ? 'red' : s.select === 2 ? 'orange' : s.hover ? 'blue' : 'lightgreen';
            context2d.fill();

            // Label
            context2d.fillStyle = s.hover ? 'white' : 'black';
            const n = this.model.segments.indexOf(s);
            context2d.fillText(String(n), xc - (n < 10 ? 4 : 8), yc + 5);
        });
    }

    drawFaces(faces) {
        const context2d = this.canvas2d.getContext('2d');
        context2d.font = '18px serif';
        context2d.strokeStyle = 'black';

        faces.forEach(f => {
            const pts = f.points;
            let [cx, cy] = [0, 0];
            context2d.beginPath();
            context2d.moveTo(pts[0].xf, -pts[0].yf);
            pts.forEach(p => {
                context2d.lineTo(p.xf, -p.yf);
                cx += p.xf;
                cy += -p.yf;
            });
            context2d.closePath();
            context2d.fillStyle = f.hover ? 'lightcyan' : f.select ? 'pink' : 'lightblue';
            context2d.fill();

            cx /= pts.length;
            cy /= pts.length;
            context2d.beginPath();
            context2d.arc(cx, cy, 12, 0, 2 * Math.PI);
            context2d.fillStyle = 'lightcyan';
            context2d.fill();
            // Label
            context2d.fillStyle = 'black';
            const n = this.model.faces.indexOf(f);
            context2d.fillText(String(n), cx - (n < 10 ? 4 : 8), cy + 5);
        });
    }

    drawModel() {
        if (this.model.faces.length === 1) { this.fit();}
        const context2d = this.canvas2d.getContext('2d');
        context2d.save();
        context2d.setTransform(1, 0, 0, 1, 0, 0);
        context2d.fillStyle = '#CCE4FF';
        context2d.fillRect(0, 0, this.canvas2d.width, this.canvas2d.height);
        context2d.restore();

        this.drawFaces(this.model.faces);
        this.drawSegments(this.model.segments);
        this.drawPoints(this.model.points);
    }
}
// 145

File: src/view2d.rs
use crate::model::Model;
use crate::helper::Helper;
use egui;
use egui::{Color32, Pos2, Stroke, Shape, Painter};

pub struct View2d {
    pub scale: f32,
    pub x_offset: f32,
    pub y_offset: f32,
}

impl View2d {
    pub fn new() -> Self {
        Self {
            scale: 1.0,
            x_offset: 0.0,
            y_offset: 0.0,
        }
    }

    pub fn fit(&mut self, model: &Model, view_rect: egui::Rect) {
        let bounds = model.get_2d_bounds();
        let model_width = bounds.x_max - bounds.x_min;
        let model_height = bounds.y_max - bounds.y_min;

        let view_width = view_rect.width();
        let view_height = view_rect.height();

        if model_width > 0.0 && model_height > 0.0 {
            self.scale = (view_width / model_width).min(view_height / model_height) / 1.2;
        } else {
            self.scale = 1.0;
        }

        self.x_offset = view_rect.center().x;
        self.y_offset = view_rect.center().y;
    }

    pub fn draw_model(&mut self, model: &Model, helper: &Helper, painter: &Painter) {
        let view_rect = painter.clip_rect();
        
        // Clear background
        painter.rect_filled(view_rect, 0.0, Color32::from_rgb(204, 228, 255));

        self.draw_faces(model, painter);
        self.draw_segments(model, painter);
        self.draw_points(model, painter);
        self.draw_helper(helper, painter);
    }

    fn draw_helper(&self, helper: &Helper, painter: &Painter) {
        if let (Some(fx), Some(fy), Some(cx), Some(cy)) = (helper.first_x, helper.first_y, helper.current_x, helper.current_y) {
            if helper.first_point.is_some() || helper.first_segment.is_some() || helper.first_face.is_some() {
                let p1 = self.to_screen(fx, fy);
                let p2 = self.to_screen(cx, cy);
                painter.line_segment([p1, p2], Stroke::new(4.0, Color32::GREEN));

                if let Some(ref label) = helper.label {
                    let circle_pos = Pos2::new(p2.x, p2.y - 16.0);
                    painter.circle_filled(circle_pos, 18.0, Color32::from_rgb(135, 206, 235)); // skyblue
                    painter.circle_stroke(circle_pos, 18.0, Stroke::new(1.0, Color32::BLACK));
                    painter.text(
                        circle_pos,
                        egui::Align2::CENTER_CENTER,
                        label,
                        egui::FontId::proportional(20.0),
                        Color32::BLACK,
                    );
                }
            }
        }
    }

    pub fn to_screen(&self, xf: f32, yf: f32) -> Pos2 {
        Pos2::new(
            self.x_offset + xf * self.scale,
            self.y_offset - yf * self.scale, // y is inverted in JS
        )
    }

    pub fn from_screen(&self, pos: Pos2) -> (f32, f32) {
        let xf = (pos.x - self.x_offset) / self.scale;
        let yf = (self.y_offset - pos.y) / self.scale;
        (xf, yf)
    }

    fn draw_faces(&self, model: &Model, painter: &Painter) {
        for (i, face) in model.faces.iter().enumerate() {
            let pts_indices = model.get_face_vertices_ccw(i);
            if pts_indices.is_empty() { continue; }

            let mut points = Vec::new();
            let mut cx = 0.0;
            let mut cy = 0.0;
            for &idx in &pts_indices {
                let p = &model.points[idx];
                let pos = self.to_screen(p.xf, p.yf);
                points.push(pos);
                cx += pos.x;
                cy += pos.y;
            }
            cx /= points.len() as f32;
            cy /= points.len() as f32;

            let color = if face.hover {
                Color32::from_rgb(224, 255, 255) // lightcyan
            } else if face.select != 0 {
                Color32::from_rgb(255, 192, 203) // pink
            } else {
                Color32::from_rgb(173, 216, 230) // lightblue
            };

            painter.add(Shape::convex_polygon(points, color, Stroke::new(1.0, Color32::BLACK)));

            // Label
            painter.circle_filled(Pos2::new(cx, cy), 12.0, Color32::from_rgb(224, 255, 255));
            painter.text(
                Pos2::new(cx, cy),
                egui::Align2::CENTER_CENTER,
                i.to_string(),
                egui::FontId::proportional(18.0),
                Color32::BLACK,
            );
        }
    }

    fn draw_segments(&self, model: &Model, painter: &Painter) {
        for (i, s) in model.segments.iter().enumerate() {
            let p1 = self.to_screen(s.p1.xf, s.p1.yf);
            let p2 = self.to_screen(s.p2.xf, s.p2.yf);
            let mid = Pos2::new((p1.x + p2.x) / 2.0, (p1.y + p2.y) / 2.0);

            let stroke_width = if s.hover { 6.0 } else { 3.0 };
            let color = match s.select {
                1 => Color32::RED,
                2 => Color32::from_rgb(255, 165, 0), // orange
                _ => if s.hover { Color32::BLUE } else { Color32::from_rgb(135, 206, 235) }, // skyblue
            };

            painter.line_segment([p1, p2], Stroke::new(stroke_width, color));

            // Arrow
            let angle = (p2.y - p1.y).atan2(p2.x - p1.x);
            let arrow_length = 20.0;
            let arrow_angle = std::f32::consts::PI / 7.0;
            let p_arrow1 = Pos2::new(
                p2.x - arrow_length * (angle + arrow_angle).cos(),
                p2.y - arrow_length * (angle + arrow_angle).sin(),
            );
            let p_arrow2 = Pos2::new(
                p2.x - arrow_length * (angle - arrow_angle).cos(),
                p2.y - arrow_length * (angle - arrow_angle).sin(),
            );
            painter.add(Shape::convex_polygon(vec![p2, p_arrow1, p_arrow2], color, Stroke::NONE));

            // Circle & Label
            let circle_color = match s.select {
                1 => Color32::RED,
                2 => Color32::from_rgb(255, 165, 0),
                _ => if s.hover { Color32::BLUE } else { Color32::from_rgb(144, 238, 144) }, // lightgreen
            };
            painter.circle_filled(mid, if s.hover { 12.0 } else { 8.0 }, circle_color);
            painter.text(
                mid,
                egui::Align2::CENTER_CENTER,
                i.to_string(),
                egui::FontId::proportional(18.0),
                if s.hover { Color32::WHITE } else { Color32::BLACK },
            );
        }
    }

    fn draw_points(&self, model: &Model, painter: &Painter) {
        for (i, p) in model.points.iter().enumerate() {
            let pos = self.to_screen(p.xf, p.yf);
            let color = match p.select {
                1 => Color32::RED,
                2 => Color32::from_rgb(255, 165, 0),
                _ => if p.hover { Color32::BLUE } else { Color32::from_rgb(135, 206, 235) },
            };

            painter.circle_filled(pos, if p.hover { 12.0 } else { 8.0 }, color);
            painter.text(
                pos,
                egui::Align2::CENTER_CENTER,
                i.to_string(),
                egui::FontId::proportional(18.0),
                if p.hover { Color32::WHITE } else { Color32::BLACK },
            );
        }
    }
}
